### PR TITLE
feat: remove `Pure` in favor of `{}`

### LINKED ISSUE #7638: Parser: strange error message
```
enum Color {
    case Red, 
}
```

```
>> Parse Error: Expected <Name> or 'Pure' before '}'

12 |     case Red, 
                 ^
                 Here
```

How can Pure be expected here?

### LINKED ISSUE #7701: Parser: replace `Pure` with `{}` and empty record with `{|}`
We have decided to make the following change:

- The syntax for the empty record `{}` should become `{|}`, meanwhile
- The syntax for the empty effect set `Pure` should become `{}`

### PR DESCRIPTION
**Pull request by herluf-ba**
Closes #7701
Closes #7638

This PR removes `Pure` as a keyword. The empty effect set is instead denoted `{}`. To avoid ambiguity the empty record type is now written `{|}`. The standard library and test have been refactored to reflect this.

Note that both `Parser` and `Parser2` has had `Pure` removed and empty record type `{|}` introduced.

This PR fixes an issue in `Parser2` as well, where effect sets could not occur in binary type expressions. An example would be `IO & {}`.

### CODE REVIEW

**Comment by herluf-ba** [comment]
@magnus-madsen Could you have a look at this when convenient?

File: main/src/ca/uwaterloo/flix/language/phase/Parser2.scala [hunk 1]
```diff
@@ -2937,17 +2938,26 @@ object Parser2 {
...
-        getItem = recordField,
-        checkForItem = NAME_FIELD.contains,
-        breakWhen = _.isRecoverType,
-        delimiterL = TokenKind.CurlyL,
-        delimiterR = TokenKind.CurlyR,
-        optionallyWith = Some((TokenKind.Bar, variableType)),
-        context = SyntacticContext.Type.OtherType
-      )
-      close(mark, TreeKind.Type.Record)
+      (nth(1), nth(2)) match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser2.scala:2941]: Can this go out of bounds?

File: main/src/library/Prelude.flix [hunk 2]
```diff
@@ -168,7 +168,7 @@ pub enum Purity3[a: Type, b: Type, c: Type, d: Type, ef: Eff] {
 /// it to be pure.
 ///
 pub def purityOf(f: a -> b \ ef): Purity[a, b, ef] = typematch f {
-    case g: a -> b \ Pure => Purity.Pure(g)
+    case g: a -> b \ {} => Purity.Pure(g)
```
Comment (magnus-madsen) [inline main/src/library/Prelude.flix:171]: This was so nicely indented before. Can we recover that? :)

Comment (magnus-madsen) [reply main/src/library/Prelude.flix:171]: Same for rest of file.

File: main/test/flix/experimental/Test.Dec.AssocType.flix [hunk 3]
```diff
@@ -43,7 +43,7 @@ mod Test.Dec.AssocType {
     }
 
     instance Iterable2[String] {
-        type Eff[String] = Pure
```
Comment (magnus-madsen) [inline main/test/flix/experimental/Test.Dec.AssocType.flix:46]: @mlutze Can we drop the experimental folder? Or at least the stuff with assoc  types?

File: main/test/flix/Test.Eff.Simplification.flix [hunk 4]
```diff
@@ -1,18 +1,24 @@
...
 
-    /// a pure function
-    def pure(): Unit \ {} = ()
+    // TODO: Uncomment with new parser
+    // Previous parser has a hard-time with effect sets in binary-type expressions.
+    // pub def testPureOrPure(): Unit \ {} & {} = pure()
+    //
+    // pub def testPureOrIo(): Unit \ {} & IO = pure()
+    //
+    // pub def testIoOrIo(): Unit \ IO & IO = io()
```
Comment (magnus-madsen) [inline main/test/flix/Test.Eff.Simplification.flix:21]: Just delete all of the above. Lines 2 to 21

File: main/src/ca/uwaterloo/flix/language/phase/Parser2.scala (hunk 1 above)
Comment (herluf-ba) [reply main/src/ca/uwaterloo/flix/language/phase/Parser2.scala:2941]: No `nth` has bound checks
```scala
 private def nth(lookahead: Int)(implicit s: State): TokenKind = {
    if (s.fuel == 0) {
      throw InternalCompilerException(s"[${currentSourceLocation()}] Parser is stuck", currentSourceLocation())
    }

    if (s.position + lookahead >= s.tokens.length - 1) {
      TokenKind.Eof
    } else {
      s.fuel -= 1
      s.tokens(s.position + lookahead).kind
    }
  }
  ```

File: main/src/library/Prelude.flix (hunk 2 above)
Comment (herluf-ba) [reply main/src/library/Prelude.flix:171]: I'm not sure what you mean? I only changed `Pure` -> `{}`. No indentation changes happened?

Comment (herluf-ba) [reply main/src/library/Prelude.flix:171]: Oh, you might mean that the `=>` align? Yep, sure I'll do that

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Thanks!

**Comment by magnus-madsen** [comment]
@stephentetley Heads up: This PR replaces `Pure` by the empty effect set written as `{ }`

**Pull request merged**